GitUI/Browsing/UserScriptMenuBuilder.cs

```diff
@@ -0,0 +1,164 @@
+                    DisplayStyle = ToolStripItemDisplayStyle.ImageAndText
+                };
+
+                button.Click += delegate
+                {
+                    if (_scriptRunner.RunScript(script.Name).NeedsGridRefresh)
+                    {
+                        _canRefreshRevisions.RefreshRevisions();
+                    }
+                };
+
+                // add to toolstrip
+                tool.Items.Add(button);
+            }
+        }
+
+        public void Build(ContextMenuStrip contextMenu)
+        {
+            var runScriptToolStripMenuItem = contextMenu.Items
+                .OfType<ToolStripMenuItem>()
+                .First(x => x.Name == RunScriptMenuItemName);
+
+            RemoveOwnScripts();
+            AddOwnScripts();
+
+            return;
+
+            void RemoveOwnScripts()
+            {
+                runScriptToolStripMenuItem.DropDown.Items.Clear();
+
+                var items = contextMenu.Items
+                    .OfType<ToolStripItem>()
+                    .Where(x => x.Name.EndsWith(OwnScriptPostfix) || x.Name == OwnScriptsSeparatorName)
+                    .ToList();
+
+                foreach (var item in items)
+                {
+                    contextMenu.Items.RemoveByKey(item.Name);
+                }
+
+                runScriptToolStripMenuItem.Visible = false;
+            }
+
+            void AddOwnScripts()
+            {
+                var lastIndex = contextMenu.Items.Count;
+                var scripts = ScriptManager.GetScripts();
+                var toRunScriptMenu = scripts.Where(x => x.Enabled)
+                    .Where(x => !x.AddToRevisionGridContextMenu)
+                    .Select(x => CreateToolStripMenuItem(x.Name, x.GetIcon()))
+                    .Cast<ToolStripItem>()
+                    .ToArray();
+
+                runScriptToolStripMenuItem.DropDown.Items.AddRange(toRunScriptMenu);
+                runScriptToolStripMenuItem.Visible = toRunScriptMenu.Any();
+
+                var toMainContextMenu = scripts.Where(x => x.Enabled)
+                    .Where(x => x.AddToRevisionGridContextMenu)
+                    .Select(x => CreateToolStripMenuItem(x.Name, x.GetIcon()))
+                    .Cast<ToolStripItem>()
+                    .ToArray();
+
+                contextMenu.Items.AddRange(toMainContextMenu);
+
+                if (toMainContextMenu.Any())
+                {
+                    contextMenu.Items.Insert(lastIndex, new ToolStripSeparator { Name = OwnScriptsSeparatorName });
+                }
+
+                return;
+
+                ToolStripMenuItem CreateToolStripMenuItem(string name, Image image)
+                {
+                    return new ToolStripMenuItem(name, image, RunScript)
+                    {
+                        Name = $"{name}{OwnScriptPostfix}"
+                    };
+                }
+
+                void RunScript(object sender, EventArgs e)
```

Comment:
I'd like to see tests for the new functionality, however local methods (especially nested ones) make testing significantly harder.
Do you think you could turn them into proper private methods? 

This way we can test them via the `TestAccessor` pattern (please search the code for examples).